File: Spacie/Shared/Models/ScanModels.swift
import Foundation

// MARK: - ScanPhase

/// Identifies the current phase of a two-phase scan.
///
/// - ``red``: Phase 1 in progress (shallow directory-only traversal).
/// - ``yellow``: Phase 1 complete, Phase 2 in progress (deep per-folder scan).
/// - ``smartGreen``: Coverage threshold reached, partial scan with virtual "Other" node.
/// - ``green``: Phase 2 complete, all data accurate.
enum ScanPhase: Sendable, Equatable {
    case red
    case yellow
    case smartGreen
    case green
}

// MARK: - ScanState

enum ScanState: Sendable, Equatable {
    case idle
    case scanning(ScanProgress)
    case completed(ScanStats)
    case cancelled
    case error(String)

    var isScanning: Bool {
        if case .scanning = self { return true }
        return false
    }

    var isCompleted: Bool {
        if case .completed = self { return true }
        return false
    }
}

// MARK: - ScanProgress

struct ScanProgress: Sendable, Equatable {
    let filesScanned: UInt64
    let directoriesScanned: UInt64
    let skippedDirectories: UInt64
    let totalLogicalSizeScanned: UInt64
    let totalPhysicalSizeScanned: UInt64
    let currentPath: String
    let elapsedTime: TimeInterval
    let estimatedTotalFiles: UInt64? // nil if unknown
    let phase: ScanPhase
    let deepScanProgress: Double?      // nil in Red, 0.0-1.0 in Yellow
    let deepScanDirsCompleted: UInt64
    let deepScanDirsTotal: UInt64
    /// Smart Scan coverage percentage (nil when not in smart scan mode or during Red phase).
    let coveragePercent: Double?
    /// Total bytes scanned so far during smart scan.
    let scannedBytes: UInt64
    /// Estimated total used space on the volume (totalSize - freeSize).
    let estimatedUsedSpace: UInt64

    init(
        filesScanned: UInt64,
        directoriesScanned: UInt64,
        skippedDirectories: UInt64,
        totalLogicalSizeScanned: UInt64,
        totalPhysicalSizeScanned: UInt64,
        currentPath: String,
        elapsedTime: TimeInterval,
        estimatedTotalFiles: UInt64?,
        phase: ScanPhase = .red,
        deepScanProgress: Double? = nil,
        deepScanDirsCompleted: UInt64 = 0,
        deepScanDirsTotal: UInt64 = 0,
        coveragePercent: Double? = nil,
        scannedBytes: UInt64 = 0,
        estimatedUsedSpace: UInt64 = 0
    ) {
        self.filesScanned = filesScanned
        self.directoriesScanned = directoriesScanned
        self.skippedDirectories = skippedDirectories
        self.totalLogicalSizeScanned = totalLogicalSizeScanned
        self.totalPhysicalSizeScanned = totalPhysicalSizeScanned
        self.currentPath = currentPath
        self.elapsedTime = elapsedTime
        self.estimatedTotalFiles = estimatedTotalFiles
        self.phase = phase
        self.deepScanProgress = deepScanProgress
        self.deepScanDirsCompleted = deepScanDirsCompleted
        self.deepScanDirsTotal = deepScanDirsTotal
        self.coveragePercent = coveragePercent
        self.scannedBytes = scannedBytes
        self.estimatedUsedSpace = estimatedUsedSpace
    }

    /// Returns the appropriate size based on the display mode.
    func totalSizeScanned(for mode: SizeMode) -> UInt64 {
        switch mode {
        case .logical: totalLogicalSizeScanned
        case .physical: totalPhysicalSizeScanned
        }
    }

    var estimatedProgress: Double? {
        guard let total = estimatedTotalFiles, total > 0 else { return nil }
        return min(1.0, Double(filesScanned) / Double(total))
    }

    var filesPerSecond: Double {
        guard elapsedTime > 0 else { return 0 }
        return Double(filesScanned) / elapsedTime
    }

    static func == (lhs: ScanProgress, rhs: ScanProgress) -> Bool {
        lhs.filesScanned == rhs.filesScanned
    }
}

// MARK: - ScanStats

struct ScanStats: Sendable, Equatable {
    let totalFiles: UInt64
    let totalDirectories: UInt64
    let totalLogicalSize: UInt64
    let totalPhysicalSize: UInt64
    let restrictedDirectories: UInt64
    let skippedDirectories: UInt64
    let scanDuration: TimeInterval
    let volumeId: String

    var filesPerSecond: Double {
        guard scanDuration > 0 else { return 0 }
        return Double(totalFiles) / scanDuration
    }
}

// MARK: - ScanEvent

enum ScanEvent: Sendable {
    case directoryEntered(path: String, depth: Int)
    case fileFound(node: RawFileNode)
    case directoryCompleted(path: String, totalSize: UInt64)
    case directorySkipped(path: String, depth: Int)
    case progress(ScanProgress)
    case restricted(path: String)
    case error(path: String, code: Int32, message: String)
    case completed(stats: ScanStats)
}

// MARK: - ScanConfiguration

struct ScanConfiguration: Sendable {
    let rootPath: URL
    let volumeId: String
    let followSymlinks: Bool
    let crossMountPoints: Bool
    let includeHidden: Bool
    let batchSize: Int
    let throttleInterval: TimeInterval // for UI updates
    let exclusionRules: ScanExclusionRules

    init(
        rootPath: URL,
        volumeId: String,
        followSymlinks: Bool = false,
        crossMountPoints: Bool = false,
        includeHidden: Bool = true,
        batchSize: Int = 1000,
        throttleInterval: TimeInterval = 0.1,
        exclusionRules: ScanExclusionRules = ScanExclusionManager.loadRules()
    ) {
        self.rootPath = rootPath
        self.volumeId = volumeId
        self.followSymlinks = followSymlinks
        self.crossMountPoints = crossMountPoints
        self.includeHidden = includeHidden
        self.batchSize = batchSize
        self.throttleInterval = throttleInterval
        self.exclusionRules = exclusionRules
    }
}

// MARK: - SizeMode

enum SizeMode: String, Sendable, CaseIterable, Identifiable {
    case logical = "logical"
    case physical = "physical"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .logical: "Logical Size"
        case .physical: "Disk Usage"
        }
    }
}

// MARK: - VisualizationMode

enum VisualizationMode: String, Sendable, CaseIterable, Identifiable {
    case sunburst = "sunburst"
    case treemap = "treemap"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .sunburst: "Sunburst"
        case .treemap: "Treemap"
        }
    }

    var systemImage: String {
        switch self {
        case .sunburst: "circle.circle"
        case .treemap: "square.grid.2x2"
        }
    }
}

// MARK: - ScanProfileType

/// Predefined scan profile that determines Tier 1 priority paths for Smart Scan.
enum ScanProfileType: String, Codable, Sendable, CaseIterable, Identifiable {
    case `default` = "default"
    case developer = "developer"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .default: "Default"
        case .developer: "Developer"
        }
    }
}

// MARK: - SmartScanSettings

/// User-configurable settings for the Smart Scan feature.
struct SmartScanSettings: Codable, Sendable, Equatable {
    /// Whether Smart Scan is enabled. When disabled, a full scan is performed.
    var isEnabled: Bool = true
    /// The active scan profile determining Tier 1 priority paths.
    var profile: ScanProfileType = .default
    /// Coverage threshold at which the scan can stop early (0.0-1.0).
    var coverageThreshold: Double = 0.95

    /// Threshold below which an automatic incremental rescan is triggered after deletion.
    /// Uses hysteresis to prevent oscillation: `max(0.80, coverageThreshold - 0.05)`.
    var rescanTriggerThreshold: Double {
        max(0.80, coverageThreshold - 0.05)
    }
}

// MARK: - SmartScanResult

/// Summary of a completed Smart Scan pass.
struct SmartScanResult: Sendable {
    /// Percentage of estimated used space that was scanned (0.0-1.0).
    let coveragePercent: Double
    /// Total bytes accounted for by deep-scanned directories.
    let scannedBytes: UInt64
    /// Estimated total used space on the volume (totalSize - freeSize).
    let estimatedUsedSpace: UInt64
    /// Number of directories that were not scanned.
    let unscannedDirectoryCount: Int
}

File: shared/src/commonMain/kotlin/com/spacie/core/model/ScanModels.kt
package com.spacie.core.model

import kotlin.experimental.ExperimentalObjCName
import kotlin.native.ObjCName

// MARK: - ScanPhase

@OptIn(ExperimentalObjCName::class)
@ObjCName("SpaScanPhase")
enum class ScanPhase {
    RED,
    YELLOW,
    SMART_GREEN,
    GREEN
}

// MARK: - ScanState

@OptIn(ExperimentalObjCName::class)
@ObjCName("SpaScanState")
sealed class ScanState {
    @ObjCName("SpaScanStateIdle")
    data object Idle : ScanState()

    @ObjCName("SpaScanStateScanning")
    data class Scanning(val progress: ScanProgress) : ScanState()

    @ObjCName("SpaScanStateCompleted")
    data class Completed(val stats: ScanStats) : ScanState()

    @ObjCName("SpaScanStateCancelled")
    data object Cancelled : ScanState()

    @ObjCName("SpaScanStateError")
    data class Error(val message: String) : ScanState()

    val isScanning: Boolean get() = this is Scanning
    val isCompleted: Boolean get() = this is Completed
}

// MARK: - ScanProgress

@OptIn(ExperimentalObjCName::class)
@ObjCName("SpaScanProgress")
data class ScanProgress(
    val filesScanned: Long,
    val directoriesScanned: Long,
    val skippedDirectories: Long,
    val totalLogicalSizeScanned: Long,
    val totalPhysicalSizeScanned: Long,
    val currentPath: String,
    val elapsedTime: Double,
    val estimatedTotalFiles: Long?,
    val phase: ScanPhase,
    val deepScanProgress: Double?,
    val deepScanDirsCompleted: Long,
    val deepScanDirsTotal: Long,
    val coveragePercent: Double?,
    val scannedBytes: Long,
    val estimatedUsedSpace: Long
) {
    fun totalSizeScanned(mode: SizeMode): Long = when (mode) {
        SizeMode.LOGICAL -> totalLogicalSizeScanned
        SizeMode.PHYSICAL -> totalPhysicalSizeScanned
    }

    val estimatedProgress: Double?
        get() {
            val total = estimatedTotalFiles ?: return null
            if (total <= 0) return null
            return minOf(1.0, filesScanned.toDouble() / total.toDouble())
        }

    val filesPerSecond: Double
        get() {
            if (elapsedTime <= 0.0) return 0.0
            return filesScanned.toDouble() / elapsedTime
        }
}

// MARK: - ScanStats

@OptIn(ExperimentalObjCName::class)
@ObjCName("SpaScanStats")
data class ScanStats(
    val totalFiles: Long,
    val totalDirectories: Long,
    val totalLogicalSize: Long,
    val totalPhysicalSize: Long,
    val restrictedDirectories: Long,
    val skippedDirectories: Long,
    val scanDuration: Double,
    val volumeId: String
) {
    val filesPerSecond: Double
        get() {
            if (scanDuration <= 0.0) return 0.0
            return totalFiles.toDouble() / scanDuration
        }
}

// MARK: - ScanEvent

@OptIn(ExperimentalObjCName::class)
@ObjCName("SpaScanEvent")
sealed class ScanEvent {
    @ObjCName("SpaScanEventDirectoryEntered")
    data class DirectoryEntered(val path: String, val depth: Int) : ScanEvent()

    @ObjCName("SpaScanEventFileFound")
    data class FileFound(val node: RawFileNode) : ScanEvent()

    @ObjCName("SpaScanEventDirectoryCompleted")
    data class DirectoryCompleted(val path: String, val totalSize: Long) : ScanEvent()

    @ObjCName("SpaScanEventDirectorySkipped")
    data class DirectorySkipped(val path: String, val depth: Int) : ScanEvent()

    @ObjCName("SpaScanEventProgress")
    data class Progress(val progress: ScanProgress) : ScanEvent()

    @ObjCName("SpaScanEventRestricted")
    data class Restricted(val path: String) : ScanEvent()

    @ObjCName("SpaScanEventError")
    data class Error(val path: String, val code: Int, val message: String) : ScanEvent()

    @ObjCName("SpaScanEventCompleted")
    data class Completed(val stats: ScanStats) : ScanEvent()
}

// MARK: - ScanConfiguration

@OptIn(ExperimentalObjCName::class)
@ObjCName("SpaScanConfiguration")
data class ScanConfiguration(
    val rootPath: String,
    val volumeId: String,
    val followSymlinks: Boolean,
    val crossMountPoints: Boolean,
    val includeHidden: Boolean,
    val batchSize: Int,
    val throttleInterval: Double,
    val exclusionBasenames: Set<String>,
    val exclusionPathPrefixes: List<String>
)

// MARK: - SizeMode

@OptIn(ExperimentalObjCName::class)
@ObjCName("SpaSizeMode")
enum class SizeMode(val id: String) {
    LOGICAL("logical"),
    PHYSICAL("physical");

    val displayName: String
        get() = when (this) {
            LOGICAL -> "Logical Size"
            PHYSICAL -> "Disk Usage"
        }
}

// MARK: - VisualizationMode

@OptIn(ExperimentalObjCName::class)
@ObjCName("SpaVisualizationMode")
enum class VisualizationMode(val id: String) {
    SUNBURST("sunburst"),
    TREEMAP("treemap");

    val displayName: String
        get() = when (this) {
            SUNBURST -> "Sunburst"
            TREEMAP -> "Treemap"
        }

    val systemImage: String
        get() = when (this) {
            SUNBURST -> "circle.circle"
            TREEMAP -> "square.grid.2x2"
        }
}

// MARK: - ScanProfileType

@OptIn(ExperimentalObjCName::class)
@ObjCName("SpaScanProfileType")
enum class ScanProfileType(val id: String) {
    DEFAULT("default"),
    DEVELOPER("developer");

    val displayName: String
        get() = when (this) {
            DEFAULT -> "Default"
            DEVELOPER -> "Developer"
        }
}

// MARK: - SmartScanSettings

@OptIn(ExperimentalObjCName::class)
@ObjCName("SpaSmartScanSettings")
data class SmartScanSettings(
    val isEnabled: Boolean,
    val profile: ScanProfileType,
    val coverageThreshold: Double
) {
    val rescanTriggerThreshold: Double
        get() = maxOf(0.80, coverageThreshold - 0.05)
}

// MARK: - SmartScanResult

@OptIn(ExperimentalObjCName::class)
@ObjCName("SpaSmartScanResult")
data class SmartScanResult(
    val coveragePercent: Double,
    val scannedBytes: Long,
    val estimatedUsedSpace: Long,
    val unscannedDirectoryCount: Int
)
